canonicalwebteam/discourse_docs/parsers.py

```diff
@@ -0,0 +1,288 @@
+# Standard library
+import re
+
+# Packages
+import dateutil.parser
+import humanize
+from bs4 import BeautifulSoup
+from jinja2 import Template
+
+# Local
+from canonicalwebteam.discourse_docs.exceptions import PathNotFoundError
+
+
+TOPIC_URL_MATCH = re.compile(
+    r"(?:/t)?(?:/(?P<slug>[^/]+))?/(?P<topic_id>\d+)(?:/\d+)?"
+)
+
+
+def resolve_path(path):
+    """
+    Given a path to a Discourse topic, resolve the path to a topic ID
+
+    A PathNotFoundError will be raised if the path is not recognised.
+
+    A RedirectFoundError will be raised if the topic should be
+    accessed at a different URL path.
+    """
+
+    topic_match = TOPIC_URL_MATCH.match(path)
+
+    if not topic_match:
+        raise PathNotFoundError(path)
+
+    topic_id = int(topic_match.groupdict()["topic_id"])
+
+    return topic_id
+
+
+def parse_index(topic):
+    """
+    Parse the index document topic to parse out:
+    - The body HTML
+    - The navigation markup
+
+    Set all as properties on the object
+    """
+
+    index = parse_topic(topic)
+    index_soup = BeautifulSoup(index["body_html"], features="html.parser")
+
+    # Get the nav
+    index["body_html"] = str(
+        get_preamble(index_soup, break_on_title="Content")
+    )
+
+    # Parse navigation
+    index["navigation"] = parse_navigation(index_soup)
+
+    return index
+
+
+def parse_topic(topic):
+    """
+    Parse a topic object from the Discourse API
+    and return document data:
+    - title: The title
+    - body_html: The HTML content of the initial topic post
+                    (with some post-processing)
+    - updated: A human-readable date, relative to now
+                (e.g. "3 days ago")
+    - forum_link: The link to the original forum post
+    """
+
+    updated_datetime = dateutil.parser.parse(
+        topic["post_stream"]["posts"][0]["updated_at"]
+    )
+
+    return {
+        "title": topic["title"],
+        "body_html": process_topic_html(
+            topic["post_stream"]["posts"][0]["cooked"]
+        ),
+        "updated": humanize.naturaltime(updated_datetime.replace(tzinfo=None)),
+        "topic_path": f"/t/{topic['slug']}/{topic['id']}",
+    }
+
+
+def parse_navigation(index_soup):
+    """
+    Given the HTML soup of a index topic
+    extract the "navigation" section
+    """
+
+    nav_soup = get_section(index_soup, "Content")
+    nav_html = "Navigation missing"
+
+    if nav_soup:
+        nav_html = str(nav_soup)
+
+    return nav_html
+
+
+def process_topic_html(html):
+    """
+    Given topic HTML, apply post-process steps
+    """
+
+    soup = BeautifulSoup(html, features="html.parser")
+    soup = replace_notifications(soup)
+    soup = replace_notes_to_editors(soup)
+
+    return str(soup)
+
+
+def replace_notes_to_editors(soup):
```

Comment:
I guess this and the next method have the biggest coupling with Discourses inner HTML representation?

Reply:
The use of `<p>` inside `<blockquote>`s is a little bit coupled to a commonmark implementation. The construction image is heavily tied to Discourse. I've tried to address this in comments now.